### ISSUE TITLE
DateTime converted to local before insertion after rails 4.2.8

### ISSUE THREAD

**Issue by jonswar** [rails/rails#36902]
### Steps to reproduce
    gem 'bundler', '1.17.3'
    require "bundler/inline"
    
    gemfile(true) do
      source "https://rubygems.org"
    
      git_source(:github) { |repo| "https://github.com/#{repo}.git" }
    
      gem "rails", '4.2.11' # Succeeds for <=4.2.7; Fails for >=4.2.8
      gem "sqlite3", '1.3.9'
    end
    
    require "active_record"
    require "minitest/autorun"
    require "logger"
    
    # This connection will do for database-independent bug reports.
    ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: "db")
    ActiveRecord::Base.logger = Logger.new(STDOUT)
    
    ActiveRecord::Base.default_timezone = :getlocal
    
    # has no effect
    #ActiveSupport.to_time_preserves_timezone = false
    #ActiveSupport.to_time_preserves_timezone = true

    ActiveRecord::Schema.define do
      create_table :things, force: true do |t|
        t.datetime :start_time
        t.datetime :start_datetime
      end
    end
    
    class Thing < ActiveRecord::Base
    end
    
    class BugTest < Minitest::Test
      def test_datetime_versus_time
        Thing.create!(start_datetime: DateTime.new(2019,1,1), start_time: Time.new(2019,1,1))
    
        start_time = Thing.connection.select_value('select start_time from things limit 1')
        start_datetime = Thing.connection.select_value('select start_datetime from things limit 1')
        puts "start_time = #{start_time}"
        puts "start_datetime = #{start_datetime}"
        assert_match /2019-01-01/, start_time
        assert_match /2019-01-01/, start_datetime
      end
    end

### Expected behavior
DateTime should be inserted in a time zone neutral way, as in rails 4.2.7 and before.

### Actual behavior
DateTime is treated as utc and converted to local time zone before insertion, resulting in a backwards incompatible change. Time still seems to work the same.

### Cause
[This commit](https://github.com/rails/rails/commit/d569f8dc5f0c6345db476ebd9f228c68c3434a29) added a `getlocal` method to `DateTime`, so now [getlocal](https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/type/date_time.rb#L15) gets called on the value.

### System configuration
**Rails version**: Succeeds for <=4.2.7; Fails for >=4.2.8

**Ruby version**: reproduced in 2.1.8, 2.3.5, 2.5.3

**Comment by jonswar** [comment]
I'm not sure if this was intentional behavior or a bug, but it seems odd that DateTime and Time are handled differently. We have a large legacy code base that uses DateTime throughout.

**Comment by rafaelfranca** [comment]
Did you changed the value of `to_time_preserves_timezone` when upgraded to the new version?

**Comment by jonswar** [comment]
Sorry, should have mentioned. Adding

    ActiveSupport.to_time_preserves_timezone = false

or

    ActiveSupport.to_time_preserves_timezone = true

after the line

    ActiveRecord::Base.default_timezone = :getlocal

does not affect the result. I also didn't see `to_time_preserves_timezone` being checked in the code path mentioned in Cause section.

**Comment by rafaelfranca** [comment]
@pixeltrix any idea about this one?

**Comment by wjessop** [comment]
@jonswar The repro script seems to behave the same for me (no failing tests) wether I use rails 4.2.7, 4.2.8 or 4.2.11. Am I missing a step?

**Comment by jonswar** [comment]
Is your server set to utc time zone? Mine is PST

**Comment by wjessop** [comment]
My local machine, BST.

**Comment by jonswar** [comment]
I get this output.

    # Running:
    
    D, [2019-08-10T20:15:29.928263 #65808] DEBUG -- :    (0.1ms)  begin transaction
    D, [2019-08-10T20:15:29.931565 #65808] DEBUG -- :   SQL (0.5ms)  INSERT INTO "things" ("start_datetime", "start_time") VALUES (?, ?)  [["start_datetime", "2018-12-31 16:00:00.000000"], ["start_time", "2019-01-01 00:00:00.000000"]]
    D, [2019-08-10T20:15:29.932764 #65808] DEBUG -- :    (0.9ms)  commit transaction
    D, [2019-08-10T20:15:29.933029 #65808] DEBUG -- :    (0.1ms)  select start_time from things limit 1
    D, [2019-08-10T20:15:29.933173 #65808] DEBUG -- :    (0.1ms)  select start_datetime from things limit 1
    start_time = 2019-01-01 00:00:00.000000
    start_datetime = 2018-12-31 16:00:00.000000
    F
    
    Finished in 0.012571s, 79.5482 runs/s, 318.1927 assertions/s.
    
      1) Failure:
    BugTest#test_datetime_versus_time [./zone.rb:49]:
    Expected /2019-01-01/ to match "2018-12-31 16:00:00.000000".
    
    1 runs, 4 assertions, 1 failures, 0 errors, 0 skips

Host and ruby information:

    jonswart> uname -a
    Darwin f45c898a2b5f.ant.amazon.com 17.7.0 Darwin Kernel Version 17.7.0: Wed Apr 24 21:17:24 PDT 2019; root:xnu-4570.71.45~1/RELEASE_X86_64 x86_64
    jonswart> ruby -v
    ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-darwin17]

I also reproduced this on linux.

**Comment by jonswar** [comment]
Here's a smaller reproduction that illustrates the cause:

    type = ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter::MysqlDateTime.new
    date = DateTime.new(2019,1,1,0,0,0)
    type.type_cast_for_database(date)

This prints "2019-01-01 00:00:00" in rails 4.2.7, and "2018-12-31 16:00:00" in rails 4.2.8.

**Comment by jonswar** [comment]
A colleague of mine independently reproduced the test failure on his ubuntu laptop.

    $ ruby -v
    ruby 2.3.1p112 (2016-04-26) [x86_64-linux-gnu]
    $ uname -a
    Linux ... #60~16.04.2-Ubuntu SMP Thu Jul 4 09:03:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**Comment by jonswar** [comment]
Is there someone else that could try to reproduce this?

**Comment by pixeltrix** [comment]
@rafaelfranca @jonswar will have a look at this

**Comment by pixeltrix** [comment]
@jonswar how are you constructing the `DateTime` objects in your application? Because based on initial investigation it seems the new behaviour is a bug fix.

As you've correctly identified the cause of the change is the addition of `getlocal` to `DateTime` which causes the following code to be executed in `ActiveRecord::Type::DateTime`:

``` ruby
zone_conversion_method = ActiveRecord::Base.default_timezone == :utc ? :getutc : :getlocal

if value.respond_to?(zone_conversion_method)
  value = value.send(zone_conversion_method)
end
```

However what's slightly confusing me is that assigning an attribute should convert it to an `ActiveSupport::TimeWithZone` instance but I'm going to ignore that for now and assume that it's because `getlocal` is called before `in_time_zone` is being called and the modularisation of the code is masking that.

So if we set the system timezone to PST, you can see an issue with `DateTime.new`:

```
>> ENV['TZ'] = 'America/Los_Angeles'
>> DateTime.new(2019,1,1,0,0,0).utc_offset
=> 0
```

As you can see, `DateTime.new` ignores the system timezone setting and this is where the error creeps in:

```
>> DateTime.new(2019,1,1,0,0,0).getlocal.to_s(:db)
=> "2018-12-31 16:00:00"
>> DateTime.new(2019,1,1,0,0,0,'-8').getlocal.to_s(:db)
=> "2019-01-01 00:00:00"
```

Note that `DateTime.now` does respect the system timezone:

```
>> DateTime.now.utc_offset
=> -25200
```

The reason that I'd consider this a bug fix is because of this:

```
>> DateTime.new(2019,1,1,0,0,0,'+2').to_s(:db)
=> "2019-01-01 00:00:00"
>> DateTime.new(2019,1,1,0,0,0,'+2').getlocal.to_s(:db)
=> "2018-12-31 14:00:00"
```

If you manually create a `DateTime` instance with a custom UTC offset then before 4.2.8 it would've been ignored and now it's respected, which seems the more correct behaviour to me.

As for an 'official' fix, Rails 4.2 is now no longer supported and I don't think we'd want to change the behaviour back anyway, so you'll need to fix this in your application. Rather than the monkey patch you've posted I'd recommend either switching to using `Time` instances or creating `DateTime` instances with the correct offsets but that will be tricky since it doesn't really do timezones so you need to know whether DST is in effect for the instance you're creating to supply the correct UTC offset. If you really need `DateTime` instances then you can use `Time.zone.local` to get a correct instance, e.g:

```
>> Time.zone.local(2019, 1, 1, 0, 0, 0).to_datetime
=> Tue, 01 Jan 2019 00:00:00 -0800
```

Sorry I don't have better news but it's probably best to do it properly now and you'll be set for upgrading to Rails 5+ then.

**Comment by jonswar** [comment]
Make sense, thanks. We will have to go through our code base and replace DateTime with Time at some point, but for right now the patch above is letting us move forward. 😬
